subgraphs/goldfinch/src/common/helpers.ts

```diff
@@ -0,0 +1,559 @@
+    entity.position = positionID;
+    entity.amount = amount;
+    entity.amountUSD = amountUSD;
+    entity.timestamp = event.block.timestamp;
+    entity.blockNumber = event.block.timestamp;
+    entity.save();
+  } else {
+    log.error("[createTransaction]Unknown transaction type {}", [
+      transactionType,
+    ]);
+    return;
+  }
+}
+
+export function updateRevenues(
+  protocol: LendingProtocol,
+  market: Market,
+  newSupplySideRevenueUSD: BigDecimal,
+  newProtocolSideRevenueUSD: BigDecimal,
+  event: ethereum.Event,
+  updateProtocol: bool = true
+  // whether to update protocol level revenues;
+  // false for interest paid to senior pool to avoid double counting revenues
+): void {
+  const newTotalRevenueUSD = newSupplySideRevenueUSD.plus(
+    newProtocolSideRevenueUSD
+  );
+  // update market's revenue
+  market.cumulativeSupplySideRevenueUSD =
+    market.cumulativeSupplySideRevenueUSD.plus(newSupplySideRevenueUSD);
+  market.cumulativeProtocolSideRevenueUSD =
+    market.cumulativeProtocolSideRevenueUSD.plus(newProtocolSideRevenueUSD);
+  market.cumulativeTotalRevenueUSD =
+    market.cumulativeTotalRevenueUSD.plus(newTotalRevenueUSD);
+  market.save();
+
+  const marketDailySnapshot = getOrCreateMarketDailySnapshot(market.id, event);
+  const marketHourlySnapshot = getOrCreateMarketHourlySnapshot(
+    market.id,
+    event
+  );
+
+  // update daily snapshot
+  marketDailySnapshot.cumulativeSupplySideRevenueUSD =
+    market.cumulativeSupplySideRevenueUSD;
+  marketDailySnapshot.cumulativeProtocolSideRevenueUSD =
+    market.cumulativeProtocolSideRevenueUSD;
+  marketDailySnapshot.cumulativeTotalRevenueUSD =
+    market.cumulativeTotalRevenueUSD;
+  marketDailySnapshot.dailySupplySideRevenueUSD =
+    marketDailySnapshot.dailySupplySideRevenueUSD.plus(newSupplySideRevenueUSD);
+  marketDailySnapshot.dailyProtocolSideRevenueUSD =
+    marketDailySnapshot.dailyProtocolSideRevenueUSD.plus(
+      newProtocolSideRevenueUSD
+    );
+  marketDailySnapshot.dailyTotalRevenueUSD =
+    marketDailySnapshot.dailyTotalRevenueUSD.plus(newTotalRevenueUSD);
+  marketDailySnapshot.save();
+
+  // update hourly snapshot
+  marketHourlySnapshot.cumulativeSupplySideRevenueUSD =
+    market.cumulativeSupplySideRevenueUSD;
+  marketHourlySnapshot.cumulativeProtocolSideRevenueUSD =
+    market.cumulativeProtocolSideRevenueUSD;
+  marketHourlySnapshot.cumulativeTotalRevenueUSD =
+    market.cumulativeTotalRevenueUSD;
+  marketHourlySnapshot.hourlySupplySideRevenueUSD =
+    marketHourlySnapshot.hourlySupplySideRevenueUSD.plus(
+      newSupplySideRevenueUSD
+    );
+  marketHourlySnapshot.hourlyProtocolSideRevenueUSD =
+    marketHourlySnapshot.hourlyProtocolSideRevenueUSD.plus(
+      newProtocolSideRevenueUSD
+    );
+  marketHourlySnapshot.hourlyTotalRevenueUSD =
+    marketHourlySnapshot.hourlyTotalRevenueUSD.plus(newTotalRevenueUSD);
+  marketHourlySnapshot.save();
+
+  // update protocol revenue
+  if (updateProtocol) {
```

Comment:
The switch is here for senior_pool to avoid double counting revenues collected from tranched pools: https://github.com/tnkrxyz/subgraphs/blob/a8198bd8ac4c278e913073b224d813651e5afbd6/subgraphs/goldfinch/src/mappings/senior_pool.ts#L387-L399